flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/ap/AffinityPropagationBulk.java

```diff
@@ -0,0 +1,449 @@
+
+			out.collect(output);
+
+		}
+	}
+
+	// Joins a Tuple3<Trg, PositiveResponsibilitiesAccumulator, SelfResponsibility> from previous step
+	// and the responsibilities. For each responsibility will calculate the availability to be sent to the
+	// responsibility source. In case of self availability will calculate the convergence too.
+	@ForwardedFieldsFirst("f0")
+	@ForwardedFieldsSecond("f0->f1; f1->f0")
+	private static class availabilityValue
+		implements JoinFunction<Tuple3<LongValue, DoubleValue, DoubleValue>,
+		Tuple3<LongValue, LongValue, DoubleValue>,
+		Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>> {
+
+		Tuple4<LongValue, LongValue, DoubleValue, DoubleValue> output = new Tuple4<>(new LongValue(), new LongValue(),
+			new DoubleValue(), new DoubleValue());
+
+		@Override
+		public Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>
+		join(Tuple3<LongValue, DoubleValue, DoubleValue> first,
+			Tuple3<LongValue, LongValue, DoubleValue> responsibility) throws Exception {
+
+			output.f0 = responsibility.f1;
+			output.f1 = responsibility.f0;
+			output.f2 = responsibility.f2;
+
+			//For self availability calculate the convergence
+			if(responsibility.f1.getValue() == responsibility.f0.getValue()){
+				output.f3 = new DoubleValue(first.f1.getValue());
+			}else{
+				//Take the responsibility value in case is positive, it will be subtracted to the positive accumulator later
+				if(responsibility.f2.getValue() > 0) {
+					output.f3 = new DoubleValue(Math.min(0, first.f1.getValue() - responsibility.f2.getValue() + first.f2.getValue()));
+				}else{
+					output.f3 = new DoubleValue(Math.min(0, first.f1.getValue() + first.f2.getValue()));
+				}
+
+			}
+
+			return output;
+		}
+	}
+
+	@ForwardedFieldsFirst("f0; f1")
+	@ForwardedFieldsSecond("f0->f1; f1->f0")
+	private static class dampedRValue
+		extends RichJoinFunction<Tuple3<LongValue, LongValue, DoubleValue>,
+		Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>, Tuple3<LongValue, LongValue, DoubleValue>> {
+
+		private double damping;
+		private double threshold;
+
+		Tuple3<LongValue, LongValue, DoubleValue> output = new Tuple3<>(new LongValue(), new LongValue(), new DoubleValue());
+
+		dampedRValue(double damping, double threshold){
+			this.damping = damping;
+			this.threshold = threshold;
+		}
+
+		@Override
+		public Tuple3<LongValue, LongValue, DoubleValue>
+		join(Tuple3<LongValue, LongValue, DoubleValue> newValue,
+			Tuple4<LongValue, LongValue, DoubleValue, DoubleValue> oldValue) {
+
+			boolean converged;
+
+			output.f0.setValue(newValue.f0.getValue());
+			output.f1.setValue(newValue.f1.getValue());
+
+			output.f2.setValue((1 - damping) * newValue.f2.getValue() + damping * oldValue.f2.getValue());
+			converged = Math.abs(newValue.f2.getValue() - oldValue.f2.getValue()) < threshold;
+
+			LongSumAggregator agg = getIterationRuntimeContext().getIterationAggregator(CONVERGENCE_MESSAGES);
+			if(converged){
+				agg.aggregate(1);
```

Comment:
Could store the value as a class member and only call `aggregate` in the close method.